File: mobile/test/data/services/gyroscope_service_test.dart
import 'dart:async';

import 'package:flutter_test/flutter_test.dart';
import 'package:sensors_plus/sensors_plus.dart';
import 'package:wheeldeck/data/services/gyroscope_service.dart';

void main() {
  group('GyroscopeService', () {
    test('integrates z angular velocity into a raw angle', () async {
      final events = StreamController<GyroscopeEvent>();
      var now = DateTime(2026, 1, 1);
      final service = GyroscopeService(
        eventStream: events.stream,
        clock: () => now,
      );

      final angles = <double>[];
      final sub = service.rawAngles.listen(angles.add);

      // First sample uses the 0.016s fallback dt. z = -2.0 (clockwise) must
      // read as positive steering.
      events.add(GyroscopeEvent(0, 0, -2.0, now));
      await Future<void>.delayed(Duration.zero);

      now = now.add(const Duration(milliseconds: 16));
      events.add(GyroscopeEvent(0, 0, 0.0, now));
      await Future<void>.delayed(Duration.zero);

      expect(angles, hasLength(2));
      expect(angles[0], closeTo(0.032, 1e-9));
      expect(angles[1], closeTo(0.032, 1e-9));

      await sub.cancel();
      await events.close();
    });

    test('clamps dt spikes and the angle range', () async {
      final events = StreamController<GyroscopeEvent>();
      var now = DateTime(2026, 1, 1);
      final service = GyroscopeService(
        eventStream: events.stream,
        clock: () => now,
      );

      final angles = <double>[];
      final sub = service.rawAngles.listen(angles.add);

      events.add(GyroscopeEvent(0, 0, -1000.0, now));
      await Future<void>.delayed(Duration.zero);

      // A 10s gap clamps dt to 0.1s; the angle clamps to ±pi.
      now = now.add(const Duration(seconds: 10));
      events.add(GyroscopeEvent(0, 0, -1000.0, now));
      await Future<void>.delayed(Duration.zero);

      expect(angles.last.abs(), closeTo(3.141592653589793, 1e-9));

      await sub.cancel();
      await events.close();
    });
  });
}

File: android/app/src/test/kotlin/dev/fazrigading/wheeldeck/data/services/GyroscopeServiceTest.kt
package dev.fazrigading.wheeldeck.data.services

import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.test.runCurrent
import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertEquals
import org.junit.Test
import kotlin.math.PI

class GyroscopeServiceTest {

    @Test
    fun `integrates z angular velocity into a raw angle`() = runTest {
        val events = MutableSharedFlow<GyroEvent>(extraBufferCapacity = 64)
        var now = 0.0
        val service = GyroscopeService(events, clock = { now })

        val angles = mutableListOf<Double>()
        val job = launch { service.rawAngles.collect { angles.add(it) } }
        runCurrent() // let the collector subscribe before emitting

        // First sample uses the 0.016s fallback dt. z = -2.0 (clockwise) must
        // read as positive steering.
        events.tryEmit(GyroEvent(z = -2.0f))
        now += 0.016
        events.tryEmit(GyroEvent(z = 0.0f))
        runCurrent()

        assertEquals(2, angles.size)
        assertEquals(0.032, angles[0], 1e-9)
        assertEquals(0.032, angles[1], 1e-9)

        job.cancel()
    }

    @Test
    fun `clamps dt spikes and the angle range`() = runTest {
        val events = MutableSharedFlow<GyroEvent>(extraBufferCapacity = 64)
        var now = 0.0
        val service = GyroscopeService(events, clock = { now })

        val angles = mutableListOf<Double>()
        val job = launch { service.rawAngles.collect { angles.add(it) } }
        runCurrent() // let the collector subscribe before emitting

        events.tryEmit(GyroEvent(z = -1000.0f))

        // A 10s gap clamps dt to 0.1s; the angle clamps to ±pi.
        now += 10.0
        events.tryEmit(GyroEvent(z = -1000.0f))
        runCurrent()

        assertEquals(PI, angles.last(), 1e-9)

        job.cancel()
    }
}
